### PR TITLE
Preserve brackets around if-lets and skip while-lets

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
r? @jieyouxu 

Tracked by #124085 

Fresh out of #129466, we have discovered 9 crates that the lint did not successfully migrate because the span of `if let` includes the surrounding brackets `(..)` like the following, which surprised me a bit.

```rust
if (if let .. { .. } else { .. }) {
// ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// the span somehow includes the surrounding brackets
}
```

There is one crate that failed the migration because some suggestion spans cross the macro expansion boundaries. Surely there is no way to patch them with `match` rewrite. To handle this case, we will instead require all spans to be tested for admissibility as suggestion spans.

Besides, there are 4 false negative cases discovered with desugared-`while let`. We don't need to lint them, because the `else` branch surely contains exactly one statement because the drop order is not changed whatsoever in this case.

```rust
while let Some(value) = droppy().get() {
..
}
// is desugared into
loop {
    if let Some(value) = droppy().get() {
        ..
    } else {
        break;
        // here can be nothing observable in this block
    } 
}
```

I believe this is the one and only false positive that I have found. I think we have finally nailed all the corner cases this time.

### CODE REVIEW

**Comment by dingxiangfei2009** [comment]
On a side note, I think the migration suggestions on unsafe attributes are broken. If we run edition migration, the `unsafe` and the surrounding brackets are misplaced, leading to code rejection. See the crate `hictor` @ `0.1.6` with hash `ccf408f4326a858c00dd845a64a86b16f360a801`.

**Comment by jieyouxu** [comment]
> There is one crate that failed the migration because some suggestion spans cross the macro expansion boundaries. Surely there is no way to patch them with match rewrite. To handle this case, we will instead require all spans to be tested for admissibility as suggestion spans.

Yes, I believe in general we want to check all spans (incl. their subspans) for `Span::from_expansion` to be `false` (modulo cases where proc-macros produce spans that have no distinguishing syntax context) before making a suggestion.

File: compiler/rustc_lint/src/if_let_rescope.rs [hunk 1]
```diff
@@ -145,7 +149,10 @@ impl IfLetRescope {
                 recovered: Recovered::No,
             }) = cond.kind
             {
-                let if_let_pat = expr.span.shrink_to_lo().between(init.span);
+                // Peel off round braces
+                let if_let_pat = source_map
+                    .span_take_while(expr.span, |&ch| ch == '(' || ch.is_whitespace())
```
Comment (jieyouxu) [inline compiler/rustc_lint/src/if_let_rescope.rs:154]: Question: does this handle weird things like `if ((((if let .. { .. } else { .. }))))` correctly? Even though I don't think this is very common...

**Review by jieyouxu — APPROVED** [review]
Thanks, this LGTM in general, I have some cases that we may want to add a test or two for (trying really hard to come up with something that might break this logic).

File: tests/ui/drop/lint-if-let-rescope.rs [hunk 2]
```diff
@@ -68,4 +68,17 @@ fn main() {
...
+        //~| WARN: this changes meaning in Rust 2024
+        //~| HELP: a `match` with a single arm can preserve the drop order up to Edition 2021
+        //~| HELP: the value is now dropped here in Edition 2024
+        // do something
+    }
+
+    while let Some(_value) = droppy().get() {
+        // Should not lint
+        break;
+    }
```
[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply tests/ui/drop/lint-if-let-rescope.rs:89]: Test added. It works and lint suggestion is applied.

File: compiler/rustc_lint/src/if_let_rescope.rs [hunk 3]
```diff
@@ -240,6 +249,22 @@ impl<'tcx> LateLintPass<'tcx> for IfLetRescope {
...
+            && let hir::ExprKind::If(cond, _conseq, _alt) = value.kind
+            && let hir::ExprKind::Let(..) = cond.kind
+        {
+            // Recall that `while let` is lowered into this:
+            // ```
+            // loop {
+            //     if let .. { body } else { break; }
+            // }
+            // ```
+            // There is no observable from the `{ break; }` block so the edition change
```
[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply compiler/rustc_lint/src/if_let_rescope.rs]: Yes. Tweaked the wording here.

**Comment by dingxiangfei2009** [comment]
@rustbot ready

- applied suggestions with additional UI test

**Comment by dingxiangfei2009** [comment]
> > There is one crate that failed the migration because some suggestion spans cross the macro expansion boundaries. Surely there is no way to patch them with match rewrite. To handle this case, we will instead require all spans to be tested for admissibility as suggestion spans.
> 
> Yes, I believe in general we want to check all spans (incl. their subspans) for `Span::from_expansion` to be `false` (modulo cases where proc-macros produce spans that have no distinguishing syntax context) before making a suggestion.

Now we are checking all the span. We bail out of emitting a suggestion because I think the fix will not be successful if any of them is not a part of a continuous user source span.

**Review by jieyouxu — APPROVED** [review]
Thanks, you can r=me after PR CI is green.

**Comment by dingxiangfei2009** [comment]
@bors r=@jieyouxu

Onto the merge train we go.

**Pull request merged**